File: mobile/ios/ShiftPlanner/ShiftPlanner/Features/Manager/Company/ViewModels/CompanyDetailsViewModel.swift
import Foundation
import Combine

@MainActor
final class CompanyDetailsViewModel: ObservableObject {
    @Published private(set) var company: AppCompany?
    @Published private(set) var branches: [AppBranchOption] = []
    @Published var branchDrafts: [CompanyBranchDraft] = []
    @Published var companyName = ""
    @Published var companyAddress = ""
    @Published var isEditing = false
    @Published var isLoading = false
    @Published var isSaving = false
    @Published var isRegeneratingInviteCode = false
    @Published var errorMessage: String?

    private let repository: CompanyRepository
    private var hasLoadedCompany = false

    init(repository: CompanyRepository? = nil) {
        self.repository = repository ?? APICompanyRepository()
    }

    var shouldShowCompanyAddressField: Bool {
        branchDrafts.isEmpty
    }

    var canSaveCompany: Bool {
        !companyName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && !isSaving
    }

    var isReadyForEditing: Bool {
        company != nil && !isLoading
    }

    func loadCompany(forceRefresh: Bool = false) async {
        guard !hasLoadedCompany || forceRefresh else { return }

        isLoading = true
        errorMessage = nil

        do {
            async let companyRequest = repository.fetchMyCompany()
            async let branchesRequest = repository.fetchBranches()

            let loadedCompany = try await companyRequest
            let loadedBranches = try await branchesRequest
            let completeCompany = loadedCompany.withBranches(loadedBranches)

            company = completeCompany
            branches = loadedBranches
            configureForm(with: completeCompany)
            hasLoadedCompany = true
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    func configureForm(with company: AppCompany) {
        guard !isEditing else { return }

        self.company = company
        companyName = company.name
        companyAddress = company.address ?? ""
        branchDrafts = company.branches.map {
            CompanyBranchDraft(
                remoteID: $0.id,
                name: $0.name,
                address: $0.address ?? ""
            )
        }
    }

    func startEditing(with company: AppCompany) {
        self.company = company
        companyName = company.name
        companyAddress = company.address ?? ""
        branchDrafts = company.branches.map {
            CompanyBranchDraft(
                remoteID: $0.id,
                name: $0.name,
                address: $0.address ?? ""
            )
        }
        errorMessage = nil
        isEditing = true
    }

    func cancelEditing(with company: AppCompany) {
        configureForm(with: company)
        errorMessage = nil
        isEditing = false
    }

    func addBranchDraft() {
        branchDrafts.append(CompanyBranchDraft())
    }

    func removeBranchDraft(id: UUID) {
        branchDrafts.removeAll { $0.id == id }
    }

    func regenerateInviteCode() async -> AppCompany? {
        guard let currentCompany = company else { return nil }

        isRegeneratingInviteCode = true
        errorMessage = nil

        do {
            let regeneratedCompany = try await repository.regenerateInviteCode()
            let updatedCompany = regeneratedCompany.withBranches(branches)
            company = updatedCompany
            configureForm(with: updatedCompany)
            isRegeneratingInviteCode = false
            return updatedCompany
        } catch {
            errorMessage = error.localizedDescription
            isRegeneratingInviteCode = false
            company = currentCompany
            return nil
        }
    }

    func saveCompanyChanges(for fallbackCompany: AppCompany) async -> AppCompany? {
        let trimmedName = companyName.trimmingCharacters(in: .whitespacesAndNewlines)
        let trimmedAddress = companyAddress.trimmingCharacters(in: .whitespacesAndNewlines)

        guard !trimmedName.isEmpty else {
            errorMessage = localized("Company name is required.", "Введите название компании.")
            return nil
        }

        if let validationError = validateBranchDrafts() {
            errorMessage = validationError
            return nil
        }

        isSaving = true
        errorMessage = nil

        do {
            let shouldUseCompanyAddress = branchDrafts.isEmpty
            var updatedCompany = try await repository.updateMyCompany(
                name: trimmedName,
                address: shouldUseCompanyAddress ? (trimmedAddress.isEmpty ? nil : trimmedAddress) : nil
            )

            let existingBranchIDs = Set(branches.map(\.id))
            let keptBranchIDs = Set(branchDrafts.compactMap(\.remoteID))
            let branchIDsToDelete = existingBranchIDs.subtracting(keptBranchIDs)

            for branchID in branchIDsToDelete {
                try await repository.deleteBranch(branchId: branchID)
            }

            var savedBranches: [AppBranchOption] = []
            for draft in branchDrafts {
                let branchName = draft.name.trimmingCharacters(in: .whitespacesAndNewlines)
                let branchAddress = draft.address.trimmingCharacters(in: .whitespacesAndNewlines)
                let normalizedAddress = branchAddress.isEmpty ? nil : branchAddress

                if let remoteID = draft.remoteID {
                    let updatedBranch = try await repository.updateBranch(
                        branchId: remoteID,
                        name: branchName,
                        address: normalizedAddress
                    )
                    savedBranches.append(updatedBranch)
                } else if !branchName.isEmpty {
                    let createdBranch = try await repository.createBranch(
                        name: branchName,
                        address: normalizedAddress
                    )
                    savedBranches.append(createdBranch)
                }
            }

            updatedCompany = updatedCompany.withBranches(savedBranches)
            company = updatedCompany
            branches = savedBranches
            companyName = updatedCompany.name
            companyAddress = updatedCompany.address ?? ""
            branchDrafts = savedBranches.map {
                CompanyBranchDraft(
                    remoteID: $0.id,
                    name: $0.name,
                    address: $0.address ?? ""
                )
            }
            isEditing = false
            isSaving = false
            hasLoadedCompany = true
            return updatedCompany
        } catch {
            errorMessage = error.localizedDescription
            isSaving = false
            company = company ?? fallbackCompany
            return nil
        }
    }

    func reset() {
        company = nil
        branches = []
        branchDrafts = []
        companyName = ""
        companyAddress = ""
        isEditing = false
        isSaving = false
        isRegeneratingInviteCode = false
        errorMessage = nil
        hasLoadedCompany = false
    }

    private func validateBranchDrafts() -> String? {
        for draft in branchDrafts {
            if draft.remoteID != nil || !draft.name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                if draft.name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    return localized("Branch name is required.", "Введите название филиала.")
                }
            }
        }

        return nil
    }
}

File: mobile/android/app/src/main/java/com/froggyriia/shiftplanner/presentation/manager/company/CompanyDetailsViewModel.kt
package com.froggyriia.shiftplanner.presentation.manager.company

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.froggyriia.shiftplanner.data.company.CompanyRepository
import com.froggyriia.shiftplanner.domain.model.WorkingHoursRange
import com.froggyriia.shiftplanner.domain.model.AppBranchOption
import com.froggyriia.shiftplanner.domain.model.AppCompany
import com.froggyriia.shiftplanner.domain.model.CompanyBranchDraft
import com.froggyriia.shiftplanner.domain.model.withBranches
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import com.froggyriia.shiftplanner.R

data class CompanyDetailsUiState(
    val company: AppCompany? = null,
    val branches: List<AppBranchOption> = emptyList(),
    val branchDrafts: List<CompanyBranchDraft> = emptyList(),
    val companyName: String = "",
    val companyAddress: String = "",
    val isEditing: Boolean = false,
    val isLoading: Boolean = false,
    val isSaving: Boolean = false,
    val isRegeneratingCode: Boolean = false,
    val errorMessage: String? = null,
    val errorMessageRes: Int? = null,
    val statusMessageRes: Int? = null
) {
    val canSave: Boolean get() = companyName.isNotBlank() && !isSaving
    val showAddressField: Boolean get() = branchDrafts.isEmpty()
}

class CompanyDetailsViewModel(
    private val repository: CompanyRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(CompanyDetailsUiState())
    val uiState: StateFlow<CompanyDetailsUiState> = _uiState.asStateFlow()

    private var hasLoaded = false

    fun loadCompany(forceRefresh: Boolean = false) {
        if (hasLoaded && !forceRefresh) return
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                val company = repository.fetchMyCompany()
                val branches = try { repository.fetchBranches() } catch (_: Throwable) { emptyList() }
                val full = company.withBranches(branches)
                applyCompany(full, branches)
                hasLoaded = true
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message, isLoading = false)
            }
        }
    }

    fun setInitialCompany(company: AppCompany) {
        if (hasLoaded) return
        applyCompany(company, company.branches)
    }

    fun startEditing() {
        val s = _uiState.value
        _uiState.value = s.copy(isEditing = true, errorMessage = null)
    }

    fun cancelEditing() {
        val company = _uiState.value.company ?: return
        applyCompany(company, _uiState.value.branches)
        _uiState.value = _uiState.value.copy(isEditing = false, errorMessage = null)
    }

    fun onNameChange(value: String) { _uiState.value = _uiState.value.copy(companyName = value) }
    fun onAddressChange(value: String) { _uiState.value = _uiState.value.copy(companyAddress = value) }
    fun clearError() {
        _uiState.value = _uiState.value.copy(errorMessage = null, errorMessageRes = null, statusMessageRes = null)
    }

    fun addBranchDraft() {
        val drafts = _uiState.value.branchDrafts + CompanyBranchDraft()
        _uiState.value = _uiState.value.copy(branchDrafts = drafts)
    }

    fun removeBranchDraft(localId: String) {
        val drafts = _uiState.value.branchDrafts.filter { it.localId != localId }
        _uiState.value = _uiState.value.copy(branchDrafts = drafts)
    }

    fun updateDraftName(localId: String, name: String) {
        val drafts = _uiState.value.branchDrafts.map {
            if (it.localId == localId) it.copy(name = name) else it
        }
        _uiState.value = _uiState.value.copy(branchDrafts = drafts)
    }

    fun updateDraftAddress(localId: String, address: String) {
        val drafts = _uiState.value.branchDrafts.map {
            if (it.localId == localId) it.copy(address = address) else it
        }
        _uiState.value = _uiState.value.copy(branchDrafts = drafts)
    }

    fun regenerateInviteCode() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isRegeneratingCode = true, errorMessage = null)
            try {
                val regenerated = repository.regenerateInviteCode()
                val branches = _uiState.value.branches
                val updated = regenerated.withBranches(branches)
                applyCompany(updated, branches)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message, isRegeneratingCode = false)
                return@launch
            }
            _uiState.value = _uiState.value.copy(isRegeneratingCode = false)
        }
    }

    fun saveChanges() {
        val state = _uiState.value
        val trimmedName = state.companyName.trim()
        if (trimmedName.isEmpty()) {
            _uiState.value = state.copy(errorMessageRes = R.string.company_name_required)
            return
        }
        for (draft in state.branchDrafts) {
            if ((draft.remoteId != null || draft.name.isNotBlank()) && draft.name.isBlank()) {
                _uiState.value = state.copy(errorMessageRes = R.string.company_branch_name_required)
                return
            }
        }

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isSaving = true, errorMessage = null)
            try {
                val useCompanyAddress = state.branchDrafts.isEmpty()
                val trimmedAddr = state.companyAddress.trim().ifEmpty { null }
                var updated = repository.updateMyCompany(
                    name = trimmedName,
                    address = if (useCompanyAddress) trimmedAddr else null
                )

                val existingIds = state.branches.map { it.id }.toSet()
                val keptIds = state.branchDrafts.mapNotNull { it.remoteId }.toSet()
                for (id in existingIds - keptIds) {
                    runCatching { repository.deleteBranch(id) }
                }

                val savedBranches = mutableListOf<AppBranchOption>()
                for (draft in state.branchDrafts) {
                    val name = draft.name.trim()
                    if (name.isEmpty()) continue
                    val addr = draft.address.trim().ifEmpty { null }
                    if (draft.remoteId != null) {
                        savedBranches += repository.updateBranch(draft.remoteId, name, addr)
                    } else {
                        savedBranches += repository.createBranch(name, addr)
                    }
                }

                updated = updated.withBranches(savedBranches)
                applyCompany(updated, savedBranches)
                _uiState.value = _uiState.value.copy(isEditing = false, isSaving = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message, isSaving = false)
            }
        }
    }

    private fun applyCompany(company: AppCompany, branches: List<AppBranchOption>) {
        _uiState.value = _uiState.value.copy(
            company = company,
            branches = branches,
            companyName = company.name,
            companyAddress = company.address ?: "",
            branchDrafts = branches.map {
                CompanyBranchDraft(
                    remoteId = it.id,
                    name = it.name,
                    address = it.address ?: ""
                )
            },
            isLoading = false,
            isSaving = false
        )
    }

    // ── Branch working hours ──────────────────────────────────────────────────

    fun loadBranchWorkingHours(
        companyId: Int,
        branchId: Int,
        onResult: (Map<Int, WorkingHoursRange>?) -> Unit
    ) {
        viewModelScope.launch {
            val hours = runCatching {
                repository.fetchBranchWorkingHours(companyId, branchId)
            }.getOrNull()
            onResult(hours)
        }
    }

    fun saveBranchWorkingHours(
        companyId: Int,
        branchId: Int,
        hours: Map<Int, WorkingHoursRange>,
        onDone: (Boolean) -> Unit
    ) {
        viewModelScope.launch {
            try {
                repository.updateBranchWorkingHours(companyId, branchId, hours)
                _uiState.value = _uiState.value.copy(statusMessageRes = R.string.company_wh_saved)
                onDone(true)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message)
                onDone(false)
            }
        }
    }
}
